### PR TITLE
Added properties for measuring and logging actor serialization times

### PR DESCRIPTION
**Pull request by bogdan-dumitrescu**

### CODE REVIEW

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java [hunk 1]
```diff
@@ -121,6 +136,10 @@ public final void run() {
             // do some trace logging
             if(this.measurement != null) {
                 log.trace(format("(%s) Message of type [%s] with id [%s] for actor [%s] took %d microsecs in queue, %d microsecs to execute, %d microsecs to serialize and %d microsecs to ack (state update %b)",this.getClass().getSimpleName(),(internalMessage != null) ? internalMessage.getPayloadClass() : "null",(internalMessage != null) ? internalMessage.getId().toString() : "null",receiverRef.getActorId(),measurement.getQueueDuration(MICROSECONDS),measurement.getExecutionDuration(MICROSECONDS),measurement.getSerializationDuration(MICROSECONDS),measurement.getAckDuration(MICROSECONDS),measurement.isSerialized()));
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java]: Now when measurements are turned on to gain insights on large state objects being serialized we always do a log.trace with a String.format. This is a unnecessary heavy operation. For now I suggest commenting it out with a TODO this this needs to be fixed later

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java [hunk 2]
```diff
@@ -121,6 +136,10 @@ public final void run() {
...
+
+                if (this.measurement.getSerializationDuration(TimeUnit.MICROSECONDS) > serializationWarnThreshold) {
+                    log.warn(format("(%s) Message of type [%s] with id [%s] for actor [%s] took %d microsecs to serialize (state update %b)",this.getClass().getSimpleName(),(internalMessage != null) ? internalMessage.getPayloadClass() : "null",(internalMessage != null) ? internalMessage.getId().toString() : "null",receiverRef.getActorId(),measurement.getSerializationDuration(MICROSECONDS),measurement.isSerialized()));
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java]: It's not the message serialization that is being measured here but the serialization of the actor state

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/app/ApplicationProtocolFactory.java [hunk 3]
```diff
@@ -37,9 +37,16 @@ public ActorLifecycleTask createHandleMessageTask(InternalActorSystem actorSyste
                                                       InternalMessage internalMessage,
                                                       PersistentActor persistentActor,
                                                       PersistentActorRepository persistentActorRepository,
-                                                      MessageHandlerEventListener messageHandlerEventListener) {
-        return new HandleMessageTask(actorSystem, receiver, receiverRef, internalMessage, persistentActor,
-                persistentActorRepository, messageHandlerEventListener);
+                                                      MessageHandlerEventListener messageHandlerEventListener,
+                                                      Boolean measure,
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/app/ApplicationProtocolFactory.java]: consider having just the serializationWarnThreshold which when null has the same semantics as measure = FALSE

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/LocalActorShard.java [hunk 4]
```diff
@@ -81,6 +83,12 @@ public LocalActorShard(PhysicalNode node,
         this.shardKey = new ShardKey(actorSystem.getName(), shard);
     }
 
+    @Autowired
+    public void setEnvironment(Environment environment) {
+        this.measure = environment.getProperty("ea.measure.shard.message.task", Boolean.class, false);
+        this.serializationWarnThreshold = environment.getProperty("ea.serialization.warn.threshold", Long.class, Long.MAX_VALUE);
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/LocalActorShard.java]: consider using null as the default which amounts to ea.measure.shard.message.task = FALSE

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/LocalActorNode.java [hunk 5]
```diff
@@ -167,8 +167,10 @@ public void handleMessage(final InternalMessage im,
                                                              receiverRef,
                                                              internalMessage,
                                                              actor,
-                                                            null,
-                                                             messageHandlerEventListener));
+                                                             null,
+                                                             messageHandlerEventListener,
+                                                             null,
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/LocalActorNode.java]: apparently the default is already NULL in this case, best to make this consistent with my other comments and remove the measure boolen altogether

**Comment by bogdan-dumitrescu** [comment]
Made suggested changes.

File: core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java [hunk 6]
```diff
@@ -135,10 +133,11 @@ public final void run() {
...
 
-                if (this.measurement.getSerializationDuration(TimeUnit.MICROSECONDS) > serializationWarnThreshold) {
-                    log.warn(format("(%s) Message of type [%s] with id [%s] for actor [%s] took %d microsecs to serialize (state update %b)",this.getClass().getSimpleName(),(internalMessage != null) ? internalMessage.getPayloadClass() : "null",(internalMessage != null) ? internalMessage.getId().toString() : "null",receiverRef.getActorId(),measurement.getSerializationDuration(MICROSECONDS),measurement.isSerialized()));
+                if (serializationWarnThreshold != null && this.measurement.getSerializationDuration(TimeUnit.MICROSECONDS) > serializationWarnThreshold) {
+                    log.warn(format("(%s) Message of type [%s] with id [%s] triggered serlalization for actor [%s] which took %d microsecs to complete",this.getClass().getSimpleName(),(internalMessage != null) ? internalMessage.getPayloadClass() : "null",(internalMessage != null) ? internalMessage.getId().toString() : "null",receiverRef.getActorId(),measurement.getSerializationDuration(MICROSECONDS)));
```
Comment (jwijgerd) [inline core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java]: there's a typo here ;-)

Comment (jwijgerd) [reply core/src/main/java/org/elasticsoftware/elasticactors/cluster/tasks/ActorLifecycleTask.java]: no worries, I'll merge and fix the typo

**Pull request merged**